src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java

```diff
@@ -342,6 +350,14 @@ public Jwt parse(String jwt) throws ExpiredJwtException, MalformedJwtException,
             //re-create the jwt part without the signature.  This is what needs to be signed for verification:
             String jwtWithoutSignature = base64UrlEncodedHeader + SEPARATOR_CHAR + base64UrlEncodedPayload;
 
+            if (signatureAlgorithm != null &&
+                    (signatureAlgorithm.compareTo(algorithm) != 0)) {
```

Comment:
Please use a `.equals` or `==` comparison, as these are idiomatic for enums.  `.compareTo` is more esoteric.